### PR TITLE
[MM][Perf][CG] Support ViT full CUDA graph for InternVL

### PR DESCRIPTION
**Pull request by oguzhankir**
## Purpose

Add ViT CUDA Graph support for InternVL models (InternVL3, InternVL2.5, InternVL2), following #38061 (Qwen3-VL). Part of #38175.

InternVL's `InternVisionModel` uses standard ViT attention with no rotary embeddings or variable-length metadata, so no extra buffer keys are needed.

## Test Plan

- Unit tests: `pytest tests/v1/cudagraph/test_encoder_cudagraph.py -v`
- Added InternVL entry to `tests/models/multimodal/generation/test_vit_cudagraph.py`
- E2E benchmark on RTX 4090

## Test Result

**Unit tests:** 36 passed ✅

**E2E Benchmark** — RTX 4090, `OpenGVLab/InternVL3-2B`, 2 images/req, 200 prompts @ 8 RPS:

| Metric | No CG | With CG | Δ |
|--------|-------|---------|---|
| Mean TTFT | 81.82 ms | 76.81 ms | ↓ 6.3% |
| Median TTFT | 74.00 ms | 72.49 ms | ↓ 2.0% |
| P99 TTFT | 147.65 ms | 122.11 ms | ↓ 17.3% |

CG config: `encoder_cudagraph_token_budgets=[256,512,1024]`, `encoder_cudagraph_max_vision_items_per_batch=4`

> Larger improvements expected with bigger models and more images per request.

## Documentation

- Added `InternVLChatModel` to `docs/design/cuda_graphs_multimodal.md`
- Added `internvl_chat` to `MODELS_SUPPORT_VIT_CUDA_GRAPH` in examples

### CODE REVIEW

**Comment by evezhier** [comment]
I've tested this and can confirm the perf. My changes are identical.

**Comment by oguzhankir** [comment]
> I've tested this and can confirm the perf. My changes are identical.

Thanks for testing and confirming! 🙏

**Comment by oguzhankir** [comment]
Gentle ping @DarkLight1337 @ywang96 🙂 Could you add the ready label so CI can run? Thanks!

**Comment by DarkLight1337** [comment]
cc @Isotr0py @shen-shanshan

**Comment by shen-shanshan** [comment]
> cc @Isotr0py @shen-shanshan

Sorry for late, I will take a look at this PR today.

File: vllm/model_executor/models/internvl.py [hunk 1]
```diff
@@ -924,3 +932,174 @@ def get_num_mm_connector_tokens(self, num_vision_tokens: int) -> int:
...
+    ) -> list[int]:
+        return [
+            n * self.num_image_token for n in self._get_internvl_patches_list(mm_kwargs)
+        ]
+
+    def get_encoder_cudagraph_per_item_input_sizes(
+        self,
+        mm_kwargs: dict[str, Any],
+    ) -> list[int]:
+        return self._get_internvl_patches_list(mm_kwargs)
```
Comment (Isotr0py) [inline vllm/model_executor/models/internvl.py]: You need to update the interface after https://github.com/vllm-project/vllm/pull/41234

Comment (oguzhankir) [reply vllm/model_executor/models/internvl.py]: Thanks for the review. Missed that #41234 landed. I'll update the interface and push shortly.

File: docs/design/cuda_graphs_multimodal.md [hunk 2]
```diff
@@ -86,6 +86,7 @@ Models opt-in to encoder CUDA Graphs by implementing the [SupportsEncoderCudaGra
 
 | Architecture | Models | CG for Image | CG for Video |
 | ------------ | ------ | ------------ | ------------ |
+| `InternVLChatModel` | `InternVL3`, `InternVL2.5`, `InternVL2` | ✅︎ | ✅︎ |
```
Comment (shen-shanshan) [inline docs/design/cuda_graphs_multimodal.md]: Does this PR support InternVL3_5?

File: tests/models/multimodal/generation/test_vit_cudagraph.py [hunk 3]
```diff
@@ -51,6 +55,16 @@ def step3_vl_chat_template(content: str) -> str:
 
 
 MODEL_CONFIGS: dict[str, VitCudagraphTestConfig] = {
+    "internvl": VitCudagraphTestConfig(
+        model="OpenGVLab/InternVL3-2B",
```
Comment (shen-shanshan) [inline tests/models/multimodal/generation/test_vit_cudagraph.py]: For CI test, we should use a model as small as possible, I suppose `OpenGVLab/InternVL3-1B` is better.

File: vllm/model_executor/models/internvl.py [hunk 4]
```diff
@@ -543,8 +544,15 @@ def _get_prompt_updates(
...
-class InternVLChatModel(nn.Module, SupportsMultiModal, SupportsPP, SupportsLoRA):
+class InternVLChatModel(
+    nn.Module,
+    SupportsMultiModal,
+    SupportsPP,
+    SupportsLoRA,
+    SupportsEncoderCudaGraph,
+):
     supports_encoder_tp_data = True
+    supports_encoder_cudagraph: ClassVar[Literal[True]] = True
```
Comment (shen-shanshan) [inline vllm/model_executor/models/internvl.py]: This `supports_encoder_cudagraph` is not needed.

File: vllm/model_executor/models/internvl.py [hunk 5]
```diff
@@ -924,3 +932,174 @@ def get_num_mm_connector_tokens(self, num_vision_tokens: int) -> int:
...
+        mm_kwargs: dict[str, Any],
+    ) -> str:
+        if "pixel_values_flat" in mm_kwargs:
+            return "image"
+        return "video"
+
+    def get_max_frames_per_video(self) -> int:
+        # InternVL has no attention-metadata buffers that depend on frame
+        # count (buffer_keys=[]), so any value is safe. Return 1.
+        return 1
```
Comment (shen-shanshan) [inline vllm/model_executor/models/internvl.py]: This method is not needed, since https://github.com/vllm-project/vllm/pull/41234/changes#diff-8214882028a01f717c0aac8d8261b29bd24e965d393b13aaa37acd96c26c6482R54-R57 has set `max_frames_per_video` default to 1.

**Review by shen-shanshan — COMMENTED** [review]
I left some comments and questions. Others LGTM.

**Comment by oguzhankir** [comment]
Thanks @shen-shanshan @Isotr0py! 

Updated:
- Migrated to get_encoder_cudagraph_item_specs per #41234 (dropped the three legacy num_items/per_item_* methods)
- Removed the redundant supports_encoder_cudagraph class attr (already provided by the Protocol) and the get_max_frames_per_video override (config default is already 1)
- Switched the CI test to OpenGVLab/InternVL3-1B (bumped max_model_len so the video case fits)
- Added InternVL3.5 to the docs table — it resolves to the same InternVLChatModel, so it's covered by this path

Verified locally on an RTX 4090: encoder cudagraph unit tests (49 passed) and the InternVL3-1B image + video e2e tests both pass. PTAL 🙏

File: docs/design/cuda_graphs_multimodal.md (hunk 2 above)
Comment (oguzhankir) [reply docs/design/cuda_graphs_multimodal.md]: Yes. InternVL3.5 resolves to the same InternVLChatModel, so it's covered. Added it to the docs table.

**Review by shen-shanshan — APPROVED** [review]
Done with my pass. Also CC @Isotr0py

**Review by Isotr0py — APPROVED** [review]

**Comment by oguzhankir** [comment]
@Isotr0py @shen-shanshan — I triggered a re-run, but the `multi-modal-models-standard-2-qwen3-plus-gemma` shard still fails deterministically on the 2 InternVL tests I added. All 8 Qwen cases in the same shard pass.

**Root cause:** Both die during encoder CUDA graph capture with a known PyTorch allocator NVML assert (`NVML_SUCCESS == r INTERNAL ASSERT FAILED at CUDACachingAllocator.cpp:1165`). 

This is highly environment-specific:
- The AMD MI325 shard passed (ROCm has no NVML path).
- It passes locally on my RTX 4090 (full GPU, no MIG).
- The new failing H200 agent is another **MIG slice** (`MIG-96b0f7c4-...`), confirming the allocator consistently fails on an NVML query during graph pool growth on MIG instances.

**Why it hits InternVL and not Qwen:** I bumped `max_model_len` to 8192 so the 16-frame video case fits. This makes the capture footprint larger than the Qwen tests (which use 4096), pushing the allocator to grow the pool at capture time, which triggers the NVML query.

I don't have a MIG device to reproduce or test a fix. I can try capping the encoder-cudagraph budget independently of `max_model_len` (so the largest video case falls back to eager), but I would need someone with an H200 to verify it. Alternatively, this might be better handled as a known PyTorch+MIG CI issue. 

How would you like to proceed?

**Comment by shen-shanshan** [comment]
@oguzhankir `input_key_by_modality` has been deprecated, you should sync the changes in https://github.com/vllm-project/vllm/pull/42288.
